app/services/account_builder.rb

```diff
@@ -0,0 +1,185 @@
+# Builds a subclassed Account object given params.
+#
+# This class supports extensions from engines via subclassing `AccountBuilder`
+# object that are named `#{AccountSubclass}Builder` (i.e. NufsAccountBuilder).
+# See `self.for()` for additional detail.
+#
+# Example usage:
+#   factory = AccountBuilder.for("NufsAccount")
+#   account = factory.new(...).build
+#
+class AccountBuilder
+
+  attr_reader :account, :account_type, :current_user, :facility, :owner_user,
+    :params, :params_key
+
+  # Initialize the instance variables.
+  #
+  # Possible `options` include:
+  #   - account_type - subclassed `Account` class name required for new/build
+  #   - account - subclassed `Account` object required for update
+  #   - facility - a facility or facility null object
+  #   - current_user - required
+  #   - owner_user - passed only for new/build account objects
+  #   - params - required
+  #
+  def initialize(options = {})
+    @account_type = valid_account_type!(options[:account], options[:account_type], options[:facility])
+    @facility = options[:facility]
+    @owner_user = options[:owner_user]
+    @current_user = options[:current_user]
+    @params_key = Account.config.account_type_to_param(@account_type)
+    @params = parse_params(options[:account], options[:params], @params_key)
+    @account = options[:account] || new_account(@account_type)
+  end
+
+  # Factory method that returns a subclassed `AccountBuilder` if one exists for
+  # the given account_type; otherwise returns `AccountBuilder` class (self).
+  # Be careful that `Object.const_get()` doesn't provide false-negatives for
+  # lazy-loaded builder classes. The `account_type` argument accepts an
+  # `Account` class object, class name string/symbol, or underscored class name
+  # string/symbol.
+  #
+  # Example usage:
+  #   AccountBuilder.for("NufsAccount") => NufsAccountBuilder
+  #   AccountBuilder.for("WithoutBuilderAccount") => AccountBuilder
+  #
+  def self.for(account_type)
+    klass = "#{account_type}_builder".classify
+    Object.const_defined?(klass) ? klass.constantize : self
```

Comment:
I see the comment about lazily-loaded classes, and that makes sense to me, but is there a specific way to go about avoiding that scenario?

Would it be better to catch a NameError (or at least I think that's the error it would raise)? That would probably handle the lazy-loading issue.
```
  klass = "#{account_type}_builder".classify
  klass.constantize
rescue NameError
  self
```